src/controller.cpp

```diff
@@ -0,0 +1,1230 @@
+    TricycleSteeringCmdVel steering_cmd = computeTricycleModelInverseKinematics(output_combined);
+    if (output_combined.linear.x < 0.0 && steering_cmd.speed > 0.0)
+    {
+      steering_cmd.speed = - steering_cmd.speed;
+      if (steering_cmd.steering_angle > 0)
+        steering_cmd.steering_angle = steering_cmd.steering_angle - M_PI;
+      else
+        steering_cmd.steering_angle = steering_cmd.steering_angle + M_PI;
+    }
+    // Apply limits to steering commands
+    steering_cmd.steering_angle = std::clamp(steering_cmd.steering_angle,
+                                  -max_steering_angle_, max_steering_angle_);
+    double steering_yaw_vel = std::clamp((steering_cmd.steering_angle - controller_state_.previous_steering_angle)
+                                          / dt.toSec(), -max_steering_yaw_vel_, max_steering_yaw_vel_);
+    double steering_angle_acc = std::clamp((steering_yaw_vel - controller_state_.previous_steering_yaw_vel)/ dt.toSec(),
+                                  -max_steering_yaw_acc_, max_steering_yaw_acc_);
+    steering_cmd.steering_angle_velocity =   controller_state_.previous_steering_yaw_vel
+                                          +  (steering_angle_acc * dt.toSec());
+    steering_cmd.steering_angle =   controller_state_.previous_steering_angle
+                                  + (steering_cmd.steering_angle_velocity * dt.toSec());
+
+    steering_cmd.speed = std::clamp(steering_cmd.speed, -max_steering_x_vel_, max_steering_x_vel_);
+    steering_cmd.acceleration = std::clamp((steering_cmd.speed - controller_state_.previous_steering_x_vel)/ dt.toSec(),
+                                  -max_steering_x_acc_, max_steering_x_acc_);
+    steering_cmd.speed =  controller_state_.previous_steering_x_vel + (steering_cmd.acceleration * dt.toSec());
+
+    controller_state_.previous_steering_angle = steering_cmd.steering_angle;
+    controller_state_.previous_steering_yaw_vel = steering_cmd.steering_angle_velocity;
+    controller_state_.previous_steering_x_vel = steering_cmd.speed;
+
+    // Compute velocities back to base_link and update controller state
+    output_steering = computeTricycleModelForwardKinematics(steering_cmd.speed, steering_cmd.steering_angle);
+    controller_state_.current_x_vel =  output_steering.linear.x;
+    controller_state_.current_yaw_vel = output_steering.angular.z;
+
+    pid_debug->steering_angle = steering_cmd.steering_angle;
+    pid_debug->steering_yaw_vel = steering_cmd.steering_angle_velocity;
+    pid_debug->steering_x_vel = steering_cmd.speed;
+
+    output_combined = output_steering;
+  }
+
+  // Publish control effort if controller enabled
+  if (!enabled_)  // Do nothing reset integral action and all filters
+  {
+    controller_state_.error_integral_lat = 0.0;
+    controller_state_.error_integral_ang = 0.0;
+  }
+
+  controller_state_.current_x_vel = new_x_vel;
+  controller_state_.current_yaw_vel = new_yaw_vel;
+  return output_combined;
+}
```

Comment:
This single function is nearly 470 lines of code. Split this up.

Reply:
Sounds good. We should probably enable sonarcloud on this repository as well. It gives indications like cyclomatic-complexity of the code. And maximum number of seperate statements in functions.

Reply:
This refactor is in progress. Should I raise an issue about CI missing?